drivers/network/netio/netio.c

```diff
@@ -0,0 +1,896 @@
+{
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+
+    SocketPut(s);
+    /* TODO: DummyCallDriver and IoComplete */
+    return STATUS_SUCCESS;
+}
+
+static struct _TRANSPORT_ADDRESS *
+TdiTransportAddressFromSocketAddress(PSOCKADDR SocketAddress)
+{
+    struct _TRANSPORT_ADDRESS *ta;
+
+    ta = ExAllocatePoolWithTag(NonPagedPool, sizeof(*ta) + sizeof(struct sockaddr), 'ADDR');
+    if (ta == NULL)
+    {
+        DbgPrint("TdiTransportAddressFromSocketAddress: Out of memory\n");
+        return NULL;
+    }
+
+    ta->TAAddressCount = 1;
+    ta->Address[0].AddressLength = sizeof(SocketAddress->sa_data);
+    ta->Address[0].AddressType = SocketAddress->sa_family;
+    memcpy(&ta->Address[0].Address[0], &SocketAddress->sa_data, ta->Address[0].AddressLength);
+
+    return ta;
+}
+
+static PTDI_CONNECTION_INFORMATION
+TdiConnectionInfoFromSocketAddress(PSOCKADDR SocketAddress)
+{
+    PTRANSPORT_ADDRESS TargetAddress;
+    PTDI_CONNECTION_INFORMATION ConnectionInformation = NULL;
+    NTSTATUS status;
+
+    TargetAddress = TdiTransportAddressFromSocketAddress(SocketAddress);
+    if (TargetAddress == NULL)
+        return NULL;
+
+    status = TdiBuildConnectionInfo(&ConnectionInformation, TargetAddress);
+    ExFreePoolWithTag(TargetAddress, 'ADDR');
+
+    if (!NT_SUCCESS(status))
+    {
+        if (ConnectionInformation != NULL)
+        {
+            ExFreePoolWithTag(ConnectionInformation, TAG_AFD_TDI_CONNECTION_INFORMATION);
+        }
+        return NULL;
+    }
+    return ConnectionInformation;
+}
+
+
+static NTSTATUS WSKAPI
+WskBind(_In_ PWSK_SOCKET Socket, _In_ PSOCKADDR LocalAddress, _Reserved_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    NTSTATUS status;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTRANSPORT_ADDRESS ta = TdiTransportAddressFromSocketAddress(LocalAddress);
+    if (ta == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    if (s->LocalAddressHandle != INVALID_HANDLE_VALUE)
+    {
+        ZwClose(s->LocalAddressHandle);
+        s->LocalAddressHandle = INVALID_HANDLE_VALUE;
+        s->LocalAddressFile = NULL;
+    }
+
+    status = TdiOpenAddressFile(&s->TdiName,
+                                ta, AFD_SHARE_REUSE, &s->LocalAddressHandle, &s->LocalAddressFile);
+
+    if (NT_SUCCESS(status))
+    {
+        memcpy(&s->LocalAddress, LocalAddress, sizeof(s->LocalAddress));
+    }
+    ExFreePoolWithTag(ta, 'ADDR');
+
+    Irp->IoStatus.Status = status;
+    return status;
+}
+
+enum direction
+{
+    DIR_SEND,
+    DIR_RECEIVE
+};
+
+static NTSTATUS WSKAPI
+WskSendTo(_In_ PWSK_SOCKET Socket,
+          _In_ PWSK_BUF Buffer,
+          _Reserved_ ULONG Flags,
+          _In_opt_ PSOCKADDR RemoteAddress,
+          _In_ ULONG ControlInfoLength,
+          _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo, _Inout_ PIRP Irp)
```

Comment:
```suggestion
static NTSTATUS WSKAPI
WskSendTo(
    _In_ PWSK_SOCKET Socket,
    _In_ PWSK_BUF Buffer,
    _Reserved_ ULONG Flags,
    _In_opt_ PSOCKADDR RemoteAddress,
    _In_ ULONG ControlInfoLength,
    _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo, _Inout_ PIRP Irp)
```
avoid depending on the func names length for indent